### PR TITLE
Add zig-build build style, river 0.1.0, rundird 0.2.0

### LINKED ISSUE #33060: Package Request: river
A dynamic tiling Wayland window manager. 

https://github.com/ifreund/river

### PR DESCRIPTION
**Pull request by ifreund**
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

Closes  #33060

### CODE REVIEW

**Comment by ericonr** [comment]
It would be nice if it could do cross :)

**Comment by travankor** [comment]
>It would be nice if it could do cross :)

blocked on #25004 ?

hopefully, not a blocker for initial packaging

**Comment by ericonr** [comment]
I don't think so. That PR is about cross compiling the compiler, not projects using zig...

And the template here isn't trying to pass any targets or anything, so the fix might even be a simple one.

**Comment by travankor** [comment]
oh woops, sorry

This is what happens when you only look at the PR title.

File: common/build-style/zig-build.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			zig_target="${XBPS_TARGET_MACHINE}-linux-gnu" zig_cpu="baseline";;
+		armv6l-musl) zig_target="arm-linux-musl" zig_cpu="generic+v6";;
+		armv6l) zig_target="arm-linux-gnu" zig_cpu="generic+v6";;
+		armv7l-musl) zig_target="arm-linux-musl" zig_cpu="generic+v7a";;
+		armv7l) zig_target="arm-linux-gnu" zig_cpu="generic+v7a";;
+		*) broken="TODO: support more target machines for the zig build style";;
+	esac
+
+	# Inform zig of the required libc include paths.
+	cat > /tmp/zig_libc.txt <<EOF
```
Comment (ericonr) [inline common/build-style/zig-build.sh:17]: Can use `<<-EOF` to allow the block to be indented.

File: common/build-style/zig-build.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	local zig_cpu
+	case $XBPS_TARGET_MACHINE in
+		aarch64-musl|i686-musl|x86_64-musl)
+			zig_target="${XBPS_TARGET_MACHINE%-musl}-linux-musl" zig_cpu="baseline";;
+		aarch64|i686|x86_64)
+			zig_target="${XBPS_TARGET_MACHINE}-linux-gnu" zig_cpu="baseline";;
+		armv6l-musl) zig_target="arm-linux-musl" zig_cpu="generic+v6";;
+		armv6l) zig_target="arm-linux-gnu" zig_cpu="generic+v6";;
+		armv7l-musl) zig_target="arm-linux-musl" zig_cpu="generic+v7a";;
+		armv7l) zig_target="arm-linux-gnu" zig_cpu="generic+v7a";;
```
Comment (ericonr) [inline common/build-style/zig-build.sh:12]: I wonder if for all archs you couldn't do something like `armv7l*)` then have a conditional for `XBPS_TARGET_LIBC` to append `-linux-gnu` or `-linux-musl`?

Would be a tad cleaner.

File: common/build-style/zig-build.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		*) broken="TODO: support more target machines for the zig build style";;
```
Comment (ericonr) [inline common/build-style/zig-build.sh:13]: Why not add ppc already? Not supported upstream yet or just not implemented?

File: common/build-style/zig-build.sh (hunk 1 above)
Comment (ericonr) [inline common/build-style/zig-build.sh:17]: Following how this is done for other build styles, should probably be a `xbps_zig_libc.txt` file in the current directory, not `/tmp`

File: srcpkgs/zig/template [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Template file for 'zig'
 pkgname=zig
 version=0.8.0
-revision=1
+revision=2
```
Comment (ericonr) [inline srcpkgs/zig/template:4]: `zig` package fixes should be a separate commit.

File: srcpkgs/rundird/template [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Template file for 'rundird'
+pkgname=rundird
+version=0.1.1
+revision=1
+build_style=zig-build
```
Comment (ericonr) [inline srcpkgs/rundird/template:5]: Why `zig-build` and not just `zig`? Because the command used is `zig build`?

File: common/build-style/zig-build.sh (hunk 3 above)
Comment (ifreund) [reply common/build-style/zig-build.sh:21]: I was just being lazy. Upstream zig has at least some level of support for everything xbps can target. Added the ppc variants to this switch and will add others when I have time though I don't have a good way to test them.

File: srcpkgs/rundird/template (hunk 5 above)
Comment (ifreund) [reply srcpkgs/rundird/template:5]: From the new commit message:

We call this "zig-build" instead of just "zig" as this build-style
relies on usage of the zig build system. In the future, other build
systems such as meson may support zig code as well. Furthermore, the
zig build system may be used to build C/C++ code as well, not just zig.

Comment (ericonr) [reply srcpkgs/rundird/template:5]: That makes sense, thanks :)

**Comment by ifreund** [comment]
2 of the patches have not yet been upstreamed, I'll be doing that shortly. As such, I'd like to hold off on merging this until they are upstreamed.

File: srcpkgs/rundird/files/rundird/run [hunk 6]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+exec /usr/bin/rundird
```
Comment (ericonr) [inline srcpkgs/rundird/files/rundird/run:2]: I don't think we spell out the full path for most services, and it shouldn't be necessary. No need for `OPTS`?

File: common/build-style/zig-build.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+	case $XBPS_TARGET_MACHINE in
+		aarch64*|i686*|x86_64*)
+			zig_target="${XBPS_TARGET_MACHINE%-musl}-linux-${zig_abi}" zig_cpu="baseline";;
+		armv6l*) zig_target="arm-linux-${zig_abi}" zig_cpu="generic+v6";;
+		armv7l*) zig_target="arm-linux-${zig_abi}" zig_cpu="generic+v7a";;
+		ppc64le*) zig_target="powerpc64le-linux-${zig_abi}" zig_cpu="baseline";;
+		ppc64*) zig_target="powerpc64-linux-${zig_abi}" zig_cpu="baseline";;
+		ppcle*) zig_target="powerpcle-linux-${zig_abi}" zig_cpu="baseline";;
+		ppc*) zig_target="powerpc-linux-${zig_abi}" zig_cpu="baseline";;
```
Comment (ericonr) [inline common/build-style/zig-build.sh:20]: I would still make this simpler, will make a diff in a bit.

**Comment by ericonr** [comment]
```
diff --git a/common/build-style/zig-build.sh b/common/build-style/zig-build.sh
index 69dea4ca55..8c11822b70 100644
--- a/common/build-style/zig-build.sh
+++ b/common/build-style/zig-build.sh
@@ -1,7 +1,5 @@
 do_build() {
-	local zig_abi
-	local zig_target
-	local zig_cpu
+	local zig_abi zig_target zig_cpu zig_arch
 
 	case $XBPS_TARGET_LIBC in
 		glibc) zig_abi="gnu";;
@@ -11,16 +9,18 @@ do_build() {
 
 	case $XBPS_TARGET_MACHINE in
 		aarch64*|i686*|x86_64*)
-			zig_target="${XBPS_TARGET_MACHINE%-musl}-linux-${zig_abi}" zig_cpu="baseline";;
+			zig_arch="${XBPS_TARGET_MACHINE%-musl}" zig_cpu="baseline";;
 		armv6l*) zig_target="arm-linux-${zig_abi}" zig_cpu="generic+v6";;
 		armv7l*) zig_target="arm-linux-${zig_abi}" zig_cpu="generic+v7a";;
 		ppc64le*) zig_target="powerpc64le-linux-${zig_abi}" zig_cpu="baseline";;
 		ppc64*) zig_target="powerpc64-linux-${zig_abi}" zig_cpu="baseline";;
 		ppcle*) zig_target="powerpcle-linux-${zig_abi}" zig_cpu="baseline";;
 		ppc*) zig_target="powerpc-linux-${zig_abi}" zig_cpu="baseline";;
-		*) broken="TODO: support more target machines for the zig build style";;
+		*) msg_error "TODO: support more target machines for the zig build style\n";;
 	esac
 
+	zig_target="${zig_arch}-linux-${zig_abi}"
+
 	# Inform zig of the required libc include paths.
 	cat > xbps_zig_libc.txt <<-EOF
 		include_dir=${XBPS_CROSS_BASE}/usr/include
```

Do for all archs what I did for x86; and broken is checked during template parsing time, not during `do_build`. This is "better" but still breaks builds; you need something in an environment file to set `broken` so XBPS doesn't error out. To avoid duplicating the logic, could create `XBPS_ZIG_ABI` ... and use those instead of local variables.

File: srcpkgs/rundird/files/rundird/run (hunk 6 above)
Comment (ifreund) [reply srcpkgs/rundird/files/rundird/run:2]: rundird doesn't have any options, I'll get rid of the unnecessary full path.

**Comment by ifreund** [comment]
Thanks for the review @ericonr! I've gone ahead and done this "properly" by defining zig target and cpu variables in the `build-profile`s and `cross-profile`s. I think this is a lot better as it makes it much easier to keep the zig target cpu features in sync with the ones used for C as the definitions are in the same file. In fact, I caught a couple of minor mistakes I had made in the previous approach this way.

This should be ready to merge from my point of view, assuming CI passes.

File: common/build-style/zig-build.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+	# TODO: This duplication between build-profiles cross-profiles is totally
+	# unnecessary. It would be nice if there was some way to avoid it.
+	if [ "$CROSS_BUILD" ]; then
+		zig_target="${XBPS_CROSS_ZIG_TARGET}"
+		zig_cpu="${XBPS_CROSS_ZIG_CPU}"
+	else
+		zig_target="${XBPS_ZIG_TARGET}"
+		zig_cpu="${XBPS_ZIG_CPU}"
+	fi
```
Comment (ericonr) [inline common/build-style/zig-build.sh:12]: What do you think about an `environment/build-style/` file that sets `archs` for the supported archs? This way you don't have to detect if these are set before `do_build` in order to trigger `broken` (or maybe it's the best way, since it avoids duplicating the information?).

Overall looks really good, just want to be sure of the best way of signaling when zig is unsupported.

Comment (ifreund) [reply common/build-style/zig-build.sh:12]: Hmm, that seems like it might be a good idea in general but I don't think we need that for zig. Zig is theoretically able to target all of xbps's targets and I've included exhaustive support for xbps's targets in this PR. 

However, support for some of the less common architectures is of course not as good and attempting to actually compile code for them may fail due to issues in the compiler and/or std. This of course depends on the exact code being compiled as well though, especially when the issues are due to stdlib deficiencies which are often easy to add temporary downstream workarounds for as the language matures.

**Comment by ifreund** [comment]
Just rebased onto master and fixed a typo in a comment that I just noticed.

**Comment by ifreund** [comment]
Zig 0.8.1 has been released with the the patches I landed to support sysroot-based cross compilation dynamically linking system libraries. Therefore we no longer need to patch zig to enable the zig-build build style and I've rebased this PR onto #32867.

**Comment by subsetpark** [comment]
@ifreund @ericonr are there still outstanding issues for this PR? I'd love to get River into my xbps managed packages...

File: common/build-style/zig-build.sh [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+	# Inform zig of the required libc include paths.
+	cat > xbps_zig_libc.txt <<-EOF
+		include_dir=${XBPS_CROSS_BASE}/usr/include
+		sys_include_dir=${XBPS_CROSS_BASE}/usr/include
+		crt_dir=${XBPS_CROSS_BASE}/usr/lib
+		msvc_lib_dir=
+		kernel32_lib_dir=
+		gcc_dir=
```
Comment (ericonr) [inline common/build-style/zig-build.sh:22]: Would it make sense to query gcc_dir from GCC itself? Would it be worth it?

File: common/cross-profiles/i686.sh [hunk 10]
```diff
@@ -8,3 +8,5 @@ XBPS_CROSS_CXXFLAGS="$XBPS_CROSS_CFLAGS"
 XBPS_CROSS_FFLAGS="$XBPS_CROSS_CFLAGS"
 XBPS_CROSS_RUSTFLAGS="--sysroot=${XBPS_CROSS_BASE}/usr"
 XBPS_CROSS_RUST_TARGET="i686-unknown-linux-gnu"
+XBPS_CROSS_ZIG_TARGET="i686-linux-gnu"
+XBPS_CROSS_ZIG_CPU="baseline"
```
Comment (ericonr) [inline common/cross-profiles/i686.sh:12]: Nit: you can enable sse2 if you want, if it isn't by default.

File: srcpkgs/zig/template (hunk 4 above)
Comment (ericonr) [reply srcpkgs/zig/template:4]: Unless I'm missing something, this wasn't resolved?

File: common/build-style/zig-build.sh (hunk 9 above)
Comment (ifreund) [reply common/build-style/zig-build.sh:22]: according to the output of `zig libc`, `gcc_dir` is only needed when targeting haiku:
```
# The directory that contains `crtbeginS.o` and `crtendS.o`
# Only needed when targeting Haiku.
gcc_dir=
```

File: srcpkgs/zig/template (hunk 4 above)
Comment (ifreund) [reply srcpkgs/zig/template:4]: This was resolved for a while but then I backport a new fix that unfortunately didn't make it into 0.8.1 and forgot about this review comment. I'll split up the commits shortly.

File: common/cross-profiles/i686.sh (hunk 10 above)
Comment (ifreund) [reply common/cross-profiles/i686.sh]: done!

**Comment by ifreund** [comment]
@ericonr Thanks for the review! I've enabled sse2 on i686 and split the zig package fix into a separate commit, this should now be good to go afaik.

File: srcpkgs/river/template [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+revision=1
+build_style=zig-build
+configure_args="$(vopt_if xwayland -Dxwayland)"
+hostmakedepends="zig pkg-config wayland-devel scdoc git"
+makedepends="wlroots-devel"
+depends="$(vopt_if xwayland xorg-server-xwayland)"
+short_desc="Dynamic tiling Wayland compositor"
+maintainer="Isaac Freund <mail@isaacfreund.com>"
+license="GPL-3.0-or-later"
+homepage="https://github.com/riverwm/river"
```
Comment (ericonr) [inline srcpkgs/river/template:13]: What about a `post_install` installing https://github.com/riverwm/river/blob/master/example/init ?

Makes it way easier to test!

**Comment by ericonr** [comment]
Ok, tested both packages, seem to be working just fine. All that's missing is installing the example file and possibly a `vdoc README.md` in rundird, what do you think?

File: srcpkgs/river/template (hunk 11 above)
Comment (ifreund) [reply srcpkgs/river/template:13]: Yeah that's definitely a good call, done.

**Comment by ifreund** [comment]
> Ok, tested both packages, seem to be working just fine. All that's missing is installing the example file and possibly a `vdoc README.md` in rundird, what do you think?

Excellent! Good call on including the example river init and rundird README.md, especially for rundird as I forgot that I never wrote a man page for that. Not that there's much that would go in a man page, the only important bit of end-user documentation is the PAM configuration.
